src/Sarif/PeekableStream.cs

```diff
@@ -0,0 +1,129 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.IO;
+
+namespace Microsoft.CodeAnalysis.Sarif.Readers
+{
+    /// <summary>
+    /// This stream provides a very limited Seekability for Streams which would benefit from an ersatz Seek(Origin)
+    /// capability, but which only need it at the beginning of a Stream.  Ie: to categorize binary versus textual 
+    /// data.
+    /// </summary>
+    internal class PeekableStream : Stream
+    {
+        private readonly Stream underlyingStream;
+        private byte[] rewindBuffer;
+        private int cursor;
+        public PeekableStream(Stream underlyingStream, int initialPeekWindow)
+        {
+            this.underlyingStream = underlyingStream;
+            this.rewindBuffer = new byte[initialPeekWindow];
+            this.cursor = 0;
+        }
+
+        public override bool CanRead => true;
+
+        // Rewind() is much weaker than Seek.  We should not advertise a capability we do not have.
+        public override bool CanSeek => false;
+
+        public override bool CanWrite => false;
+
+        public override long Length => underlyingStream.Length;
+
+        public override long Position 
+        {
+            get => Math.Min(underlyingStream.Position, cursor);
+            set => throw new System.NotImplementedException();
+        }
+
+        public override void Flush()
+        {
+            throw new System.NotImplementedException();
+        }
+
+        public override int Read(byte[] buffer, int offset, int count)
```

Comment:
>Read [](http://example.com/codeflow?start=28&length=4)

ok! I think I understand. your peekable stream is kinda like a pocket knife with a 'stop' before it fully opens. users can read the header bytes and you will 'stop' when you hit the transition point that will require you to abandon the rewind buffer. this gives the caller the change to easily fully inspect the header without worrying about an over-read. At this point, either the caller presumably rewinds (to get the entire stream by passing it to another reader) *OR* moves forward.

Two things: 

1) this is weird, but for readability, I might have ordered this code to reflect these three states reflecting where we are in the stream, 1) within the rewind buffer, 2) at the transition point, 3) past the transition point.

Simple because this makes it easier to build a mental model of what's going on by reading your commented code. 

Second thing, consider a large comment at the top that describes the general approach. For fun, ask Copilot to generate it. :) haha.

Reply:
The "stop" point is more an implementation detail to make implementation easier than an API feature.  The caller is in control of the buffer size.  That's 100% of what they need to know to use the class correctly.  I would not want them forming dependencies on the partial read behavior.

What the partial reads do, is let me operate over one byte range per Read() call.  That simplifies things a bit.  It would be harder to read and get correct (I claim) if I got hung up on delivering all the requested bytes, and had to compose both an internal buffer read and a call into the underlying stream, in the same call.